airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/AbstractDestination.java

```diff
@@ -0,0 +1,162 @@
+import io.airbyte.protocol.models.ConfiguredAirbyteCatalog;
+import io.airbyte.protocol.models.ConfiguredAirbyteStream;
+import io.airbyte.protocol.models.SyncMode;
+import io.airbyte.queue.BigQueue;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+public abstract class AbstractDestination implements Destination, DestinationWriteOperations {
+
+  /**
+   * Strategy:
+   * <p>
+   * 1. Create a temporary table for each stream
+   * </p>
+   * <p>
+   * 2. Accumulate records in a buffer. One buffer per stream.
+   * </p>
+   * <p>
+   * 3. As records accumulate write them in batch to the database. We set a minimum numbers of records
+   * before writing to avoid wasteful record-wise writes.
+   * </p>
+   * <p>
+   * 4. Once all records have been written to buffer, flush the buffer and write any remaining records
+   * to the database (regardless of how few are left).
+   * </p>
+   * <p>
+   * 5. In a single transaction, delete the target tables if they exist and rename the temp tables to
+   * the final table name.
+   * </p>
+   *
+   * @param config - integration-specific configuration object as json. e.g. { "username": "airbyte",
+   *        "password": "super secure" }
+   * @param catalog - schema of the incoming messages.
+   * @return consumer that writes singer messages to the database.
+   * @throws Exception - anything could happen!
+   */
+  @Override
+  public DestinationConsumer<AirbyteMessage> write(JsonNode config, ConfiguredAirbyteCatalog catalog) throws Exception {
+    connectDatabase(config);
+    final Map<String, BufferedWriteConfig> writeBuffers = new HashMap<>();
+    final Set<String> schemaSet = new HashSet<>();
+    // create tmp tables if not exist
+    for (final ConfiguredAirbyteStream stream : catalog.getStreams()) {
+      final String streamName = stream.getStream().getName();
+      final String schemaName = getNamingResolver().getIdentifier(getDefaultSchemaName(config));
+      final String tableName = getNamingResolver().getRawTableName(streamName);
+      final String tmpTableName = getNamingResolver().getTmpTableName(streamName);
+      if (!schemaSet.contains(schemaName)) {
+        createSchema(schemaName);
+        schemaSet.add(schemaName);
+      }
+      createTable(schemaName, tmpTableName);
+
+      final Path queueRoot = Files.createTempDirectory("queues");
+      final BigQueue writeBuffer = new BigQueue(queueRoot.resolve(streamName), streamName);
+      final SyncMode syncMode = stream.getSyncMode() == null ? SyncMode.FULL_REFRESH : stream.getSyncMode();
+      writeBuffers.put(streamName, new BufferedWriteConfig(schemaName, tableName, tmpTableName, writeBuffer, syncMode));
+    }
+    // write to tmp tables
+    // if success copy delete main table if exists. rename tmp tables to real tables.
+    return new BufferedRecordConsumer(this, writeBuffers, catalog);
```

Comment:
can we avoid passing this here? it ends up being effectively a circular dependency.